apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/Interval.cs

```diff
@@ -0,0 +1,381 @@
+                case 'Y':
+                    state.MayBeTerminal = true;
+                    state.IsTerminal = false;
+                    state.Years = ParseInt();
+                    state.NextAllowed = IntervalParsingState.AfterY;
+                    break;
+                case 'M' when !state.IsTime:
+                    state.MayBeTerminal = true;
+                    state.IsTerminal = false;
+                    state.Months = ParseInt();
+                    state.NextAllowed = IntervalParsingState.AfterMonth;
+                    break;
+                case 'D':
+                    state.MayBeTerminal = true;
+                    state.IsTerminal = false;
+                    state.Days = ParseInt();
+                    state.NextAllowed = IntervalParsingState.AfterD;
+                    break;
+                case 'T':
+                    state.MayBeTerminal = false;
+                    state.IsTerminal = false;
+                    state.IsTime = true;
+                    state.NextAllowed = IntervalParsingState.AfterT;
+                    break;
+                case 'H':
+                    state.MayBeTerminal = true;
+                    state.IsTerminal = false;
+                    state.Hours = ParseInt();
+                    state.NextAllowed = IntervalParsingState.AfterH;
+                    break;
+                case 'M' when state.IsTime:
+                    state.MayBeTerminal = true;
+                    state.IsTerminal = false;
+                    state.Minutes = ParseLong();
+                    state.NextAllowed = IntervalParsingState.AfterMins;
+                    break;
+                case 'S':
+                    state.MayBeTerminal = true;
+                    state.IsTerminal = true;
+                    state.Seconds = ParseSeconds();
+                    state.NextAllowed = null;
+                    break;
+                default:
+                    throw new InvalidOperationException("There's a bug in Interval.Parse.");
+            }
+
+            state.Start = end + 1;
+
+        } while (state.Start < text.Length && !state.IsTerminal);
+
+        // We are at a terminal state but we haven't parsed the whole string yet.
+        if (state.IsTerminal && state.Start < text.Length)
+        {
+            throw new FormatException($"Can't parse {text.Substring(state.Start)} after reaching the end of a valid duration.");
+        }
+
+        // We parsed the whole string but we ended up at a state that's not terminal.
+        if (!state.MayBeTerminal)
+        {
+            throw new FormatException($"{text} ending is not correctly formatted.");
+        }
+
+        int totalMonths =
+            state.Years * MonthsInAYear
+            + state.Months;
+
+        BigInteger totalNanoseconds =
+            state.Hours * (BigInteger)NanosecondsInAnHour
+            + state.Minutes * (BigInteger)NanosecondsInAMinute
+            + SecondsToNanoseconds(state.Seconds);
```

Comment:
This will throw ArgumentOutOfRangeException rather than FormatException if more than 9 decimal places have been specified.